### ISSUE TITLE
Non-object method calls become sugar for call of lang.T function 

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/ballerina-spec#111]
We are currently planning for the standard library to have a lang.T module for each built-in type T, which would contain functions relevant to type T.

This raises the question of what the's real difference between a function in a lang.T module and a built-in method on T. As @sanjiva observed, from the perspective of the average user, if, for example, str is an alias for lang.string, and s is variable of type string, there's not a lot of difference between `str:foo(s)` and `s.foo()`.

This issue proposes to eliminate (or at least reduce) the difference. Thus the semantics of a method call `x.foo(args)` would be

- if the basic type of x is object, then call the foo object of x
- otherwise, if the basic type of x is T, then call `m:foo(x, args)` where m is an automatic module alias corresponding to lang.T

The main advantage of this is that it avoids the need to divide up functionality between built-in methods and library functions, when there does not seem to be any obvious right way to make the division. There are multiple criteria that might be used. For example:

- is there a conventional notation that is either functional or method-like (e.g. sqrt(x) rather than x.sqrt())?
- is the function frequently used?
- is the function primitive (not implementable in terms of other functions)?
- does the function need special typing rules (which cannot yet be expressed in the language), e.g related to parameterized typing?
- can the function be implemented in Ballerina or does it need to be built-in to the compiler?
- how stable is the functionality?

It's hard to see how to resolve these conflicts so that it will be obvious to a user whether something is a function or a method (rather than being something they just have to remember), as well as maximally convenient.

There are few other advantages:

- A function in a module for lang.T may have a first argument that is a _subtype_ of T, whereas for a built-in method on type T, there is no mechanism for the self type to be anything other than T. For example, we could have a list:join(string[]) function.
- These functions can easily be implemented in Ballerina if we want (very useful for all the string functions)

It does raise some issues that will need to be resolved:

- how would this impact the lang spec? how would these lang.* modules be specfiied?
- what would we do about functions whose type cannot yet be written in the language?
    - devise some ad-hoc annotation
    - retain the idea that a very few special methods are built-in, and that a method call first looks at these built-in methods and then translates it to a library call
- how exactly does static typing work, when the self type is a union? (the current spec is not clear on this either)

**Comment by jclark** [comment]
Note that not every function in lang.T would be invokable as a method: only those whose first argument was a subtype of T. We could also require that functions explicitly indicate (e.g. via an annotation) that they could be invoked via the built-in method syntax.

I think this will work out nicely for string conversion. Typically functions come in pairs: doing a particular kind of conversion to a string, and doing the reverse. We can have e.g.

```
int:toString(int n) returns string;
int:fromString(string s) returns int|error;
```

These are nicely parallel (as they should be), but the first can still be invoked using convenient method syntax.

**Comment by sanjiva** [comment]
Big +1 for this approach.

Something that irked me for a while:
```int:fromString(string s) returns int|error```
or
```string:toInt(string s) returns int|error```
This way `myInt.toString()` works as does `myString.toInt()`. 

Of course we can choose to* or from* and do it either way.

**Comment by Shan1024** [comment]
On a side note, this will enable to implement `int:min()`, `int:max()`, etc which was missing so far and could not fit into anywhere else.

**Comment by jclark** [comment]
Here's another interesting twist. The semantics for x.foo() when x is an object could be:

1. If x has a method foo, then call foo
2. Otherwise, call object:foo(x)

This would allow objects to in effect have built-in methods on objects. Thus for example we could define object:map to work in terms of the __iterator method.

**Comment by jclark** [comment]
One problem is we need a way to use annotations to write the types of of parameterized functions. Here's an idea for this

```
// module list
// The @typeParam annotation says that
// when T is used in a function declaration, it behaves like a type param;
// the declared type is the upper bound on the param.
@typeParam
private type T any|error;
// Use like a normal type name
function map(T[], function (T arg) returns R fun) returns R[];

// Example of constraining type bound
// module error
@typeParam
private type R record {};
@typeParam
private type S string; 
function detail(error<S,R>) returns R;
```

**Comment by sanjiva** [comment]
Ref invoking object methods via this approach:

Say I have objects A & B and I've been happily using a.map() etc. to get functional iteration on an instance of those objects. If I write another object C and maybe in that object map() is a natural function for it and I write that.

Then I would implicitly "hide" the object:map() functionality and that could confuse mort.

**Comment by jclark** [comment]
We agreed the following desugaring semantics. `E.foo(args)` desugars based on the static type T of E as follows:

- if T is a subtype of object {}, then it is treated as a method call on an object
- if T is a subtype of B for for some basic type B that is not nil or object, then it desugars to `M:foo(E, args`) where M comes from B
- if T is a subtype of B|() for some basic type non-nil basic type B, then it desugars to `E != () ? E.foo(args) : ()`, where the `E.foo(args)` is recursively desugared (as above)using T without nil, and E is only evaluated once
- otherwise (T must contain values from two or more basic types other than nil) it desugars to `any:foo(E, args)` where `any` is a module-name TBD

Note that nil lifting is not done in the last case. There is no need to and it make `any.toString()` not work.

**Comment by jclark** [comment]
Semantics of a method call expression `E.foo(args)` where expression E has compile-time type T is as follows:

1. if T is not () and T is a subtype of a type `B?` where B is basic type that is not nil and not object, and there is a function M:foo, where M is the module for B, then
    1. if nil is in T, then this is first turned turns into E == () ? () : E.foo(args), except that E is evaluated only once
    2. E.foo(args) turns into M:foo(E, args)
2. if T is not () and T is a subtype of a type Obj?, where Obj is a subtype of object{}, and Obj has a method foo, then
    1. (same as 1a)
    2. E.foo(args) calls method foo on the result of evaluating E with arglist args
3. otherwise (either T is nil or T contains values from multiple basic types where at least two are non-nil), if there is a method any:foo, then
    1. E:foo(args) turns into any:foo(E, args)
4. otherwise, compile-time error

The part that remains to work out is nil lifting #133.

**Comment by jclark** [comment]
I have made the fallback to lang.object a separate issue #144, which we can deal with later.

I have written this up without nil lifting, given that the spec has never defined nil lifting for method call expressions.
